### PR TITLE
Add TerminateOnNan handler with tests and docs

### PR DESCRIPTION
**Pull request by vfdev-5**
Addresses #201 

Following the discussion from #219 we opt to an optional single tensor value checking solution.

### CODE REVIEW

File: ignite/handlers/terminate_on_nan.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import logging
+import numbers
+
+
+class TerminateOnNan(object):
+    """TerminateOnNan handler can be used to stop the training if the `process_function`'s output becomes
+    NaN or infinite `torch.tensor` or number.
+
+    Args:
+        output_transform (Callable, optional): a callable that is used to transform the
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/handlers/terminate_on_nan.py]: I thought about the case of our supervised trainer which outputs [loss.item()](https://github.com/pytorch/ignite/blob/master/ignite/engine/__init__.py#L40). 

But I agree that checking a collection of var size tensors can be interesting...
Okay, let me change it

[the remark this answers is not in the document]
Comment (alykhantejani) [reply ignite/handlers/terminate_on_nan.py]: thanks

**Comment by vfdev-5** [comment]
[WIP] I'm trying to recode the handler such that the output can be of any type and we check if it contains inf/nan tensor/number.

At the moment, as based on `apply_on_tensor` I can not handle the case of with collection of numbers.

**Comment by vfdev-5** [comment]
Refactored `apply_to_tensor` into `apply_to_type` method and use it to handle cases of number/tensor. 

@alykhantejani could you please review the changes
